core/domain/classifier_services.py

```diff
@@ -424,41 +375,239 @@ def delete_classifier_training_job(job_id):
+    if training_job_exploration_mapping_model:
+        return classifier_domain.TrainingJobExplorationMapping(
+            training_job_exploration_mapping_model.exp_id,
+            training_job_exploration_mapping_model.exp_version,
+            training_job_exploration_mapping_model.state_name,
+            training_job_exploration_mapping_model.algorithm_id_to_job_id_map)
+
+
+def migrate_exploration_training_job(training_job_exploration_mapping):
+    """Migrate exploration training job to latest version of algorithm_id
+    and algorithm_version.
+
+    This function lazily migrates an older classifier training job and
+    trains new classifiers. Specifically, if training job exploration mapping of
+    an <exploration, verison, state> triplet is missing job_id for some
+    algorithm_id, or if the job_id exists but it has been trained on a now
+    obsolete algorithm, we re-submit the jobs.
+
+    The function goes through existing training job exploration mapping and
+    identifies three different types of algorithm IDs.
+        1. algorithm_ids_to_upgrade: Those which exist but needs to be
+            upgraded a new algorithm (because existing one has been deprecated)
+            by re-submitting the training job.
+        2. algorithm_ids_to_add: Those which doesn't exist and needs to be added
+            by submitting a new training job.
+        3. algorithm_ids_to_remove: Those which needs to be removed since these
+            algorithms are no longer supported.
+
+    Once all three types of algorithm IDs are filtered, the function performs
+    specific tasks tailored to each of them. Its called lazy migration because
+    it happens only when there is a query to retrieve trained model for given
+    <exploration, version, state> and algorithm_id.
+
+    Args:
+        training_job_exploration_mapping: TrainingJobExplorationMapping. Domain
+            object containing exploration to training job id mapping. This
+            mapping is used to figure out jobs that needs to be re-submitted /
+            added or removed.
+    """
+    exp_id = training_job_exploration_mapping.exp_id
+    exp_version = training_job_exploration_mapping.exp_version
+    state_name = training_job_exploration_mapping.state_name
+
+    exploration = exp_fetchers.get_exploration_by_id(
+        exp_id, version=exp_version)
+    interaction_id = exploration.states[state_name].interaction.id
+
+    algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+        interaction_id]['algorithm_id']
+    algorithm_version = feconf.INTERACTION_CLASSIFIER_MAPPING[
+        interaction_id]['algorithm_version']
+
+    algorithm_id_to_algorithm_version = {
+        algorithm_id: algorithm_version
+    }
+    possible_algorithm_ids = [algorithm_id]
+
+    algorithm_ids_to_add = [
+        algorithm_id for algorithm_id in possible_algorithm_ids
+        if not (
+            training_job_exploration_mapping.algorithm_id_to_job_id_map.
```

Comment:
I find myself somewhat confused about the "algorithm_id_to_job_id_map". Surely multiple jobs could be based on the same algorithm, conceptually?

Reply:
Solved.